### DISCUSSION TITLE
Answered:  Long Running, stuck queries, in Digital Ocean (wrong pooling mode) + more insights

### DISCUSSION THREAD

**Discussion by matteius** [saleor/saleor#10021] [Deployment] [answered]
What is the difference between 3.3.x and 3.4.x as the releases are both happening?

I am trying to find a stable point in time that my store used saleor and every version I deploy has long running queries and sporadic graphQL unavailability, but for a few weeks I had a working version in prod that never had outages.   I keep bouncing between 3.3.x and 3.4.x without luck, but maybe it was really upgrading react-storefront that bit me and I never noticed it, I am not sure.   Lots of long running queries (hour or more) chewing up resources is all I can figure out so far, because it's sporadically online and fast to load things and then sometimes the resources are super slow to load a product, and then sometimes it just errors out and gets nothing from the graphQL API.  Any help would be appreciated figuring out a stable point in time to roll back or forward to.

Like here is a current state where it happened to get the menu data but then failed to fetch any product details, but in a side bar graphQL playground I got the latest 20 product details (though sometimes that borks too):
![image](https://user-images.githubusercontent.com/479892/174462797-e9f62f7f-d69b-4288-9021-317ff64945d6.png)

Current long running queries:
![image](https://user-images.githubusercontent.com/479892/174462844-582cf21d-899f-47de-b6f8-a1a5d240d888.png)

```
SELECT COUNT(*) FROM (SELECT DISTINCT "attribute_attribute"."id" AS Col1, "attribute_attribute"."private_metadata" AS Col2, "attribute_attribute"."metadata" AS Col3, "attribute_attribute"."slug" AS Col4, "attribute_attribute"."name" AS Col5, "attribute_attribute"."type" AS Col6, "attribute_attribute"."input_type" AS Col7, "attribute_attribute"."entity_type" AS Col8, "attribute_attribute"."unit" AS Col9, "attribute_attribute"."value_required" AS Col10, "attribute_attribute"."is_variant_only" AS Col11, "attribute_attribute"."visible_in_storefront" AS Col12, "attribute_attribute"."filterable_in_storefront" AS Col13, "attribute_attribute"."filterable_in_dashboard" AS Col14, "attribute_attribute"."storefront_search_position" AS Col15, "attribute_attribute"."available_in_grid" AS Col16 FROM "attribute_attribute" LEFT OUTER JOIN "attribute_attributeproduct" ON ("attribute_attribute"."id" = "attribute_attributeproduct"."attribute_id") LEFT OUTER JOIN "attribute_attributevariant" ON ("attribute_attribute"."id" = "att
```
Later on there are more types of queries that need to be axed:
![image](https://user-images.githubusercontent.com/479892/174463224-c93d7404-f68c-4463-9204-1713e75af865.png)

I increased workers from 4 to 8 (already have 2 nodes) and I see now twice as many long running queries at the time that the site gets wedged. I am now 100% confident that long running queries are hosing me, but why they are long running make no sense.  I even upgraded my DB instance from 1 GB to 2 GB of ram so its way overpowered now for what it does to no avail.

![image](https://user-images.githubusercontent.com/479892/174465560-08d5a701-3dd6-4a14-91d9-f93e8d794920.png)

**Comment by matteius** [comment]
3.4.4 is deployed now and is working for the moment.  It works more so on desktop, on mobile it was only working when I turned my wifi off and my Dad is having the same experience in a different state, though for me on 3.4.4 turning wifi on also just worked  There are still a lot of questions to be had around slow queries.  For example I added a combined index to help this query before realizing that the table is empty and it shouldn't be slow at all:
```
SELECT "product_collectiontranslation"."id", "product_collectiontranslation"."language_code", "product_collectiontranslation"."seo_title", "product_collectiontranslation"."seo_description", "product_collectiontranslation"."collection_id", "product_collectiontranslation"."name", "product_collectiontranslation"."description" FROM "product_collectiontranslation" WHERE ("product_collectiontranslation"."collection_id" IN (11, 4, 7, 13, 15, 14, 2) AND "product_collectiontranslation"."language_code" IN ('en-us'))
```

**Reply by matteius** [reply]
Is there a way to disable translations?

**Comment by timuric** [comment, marked as answer]
@matteius you would have to narrow down performance issues with a specific API query and provide more details about your environment. We haven't experienced any issues in our cloud environment, so this could be quite specific to your configuration.
My advice would be to use a free cloud version of Saleor for development, to free up yourself from the DevOps burden.

**Reply by matteius** [reply]
> you would have to narrow down performance issues with a specific API query and provide more details about your environment.

@timuric Problem is there are many API queries getting wedged, not just a specific one.

Here are some from yesterday that I sent along to DO:

```
Some of the queries have been copied and pasted below, This first query has been running for almost 2 hours!  114.60095 minutes currently
CLOSE "_django_curs_140695772387072_sync_12"

#6832135ms
SELECT COUNT(*) FROM (SELECT DISTINCT "attribute_attribute"."id" AS Col1, "attribute_attribute"."private_metadata" AS Col2, "attribute_attribute"."metadata" AS Col3, "attribute_attribute"."slug" AS Col4, "attribute_attribute"."name" AS Col5, "attribute_attribute"."type" AS Col6, "attribute_attribute"."input_type" AS Col7, "attribute_attribute"."entity_type" AS Col8, "attribute_attribute"."unit" AS Col9, "attribute_attribute"."value_required" AS Col10, "attribute_attribute"."is_variant_only" AS Col11, "attribute_attribute"."visible_in_storefront" AS Col12, "attribute_attribute"."filterable_in_storefront" AS Col13, "attribute_attribute"."filterable_in_dashboard" AS Col14, "attribute_attribute"."storefront_search_position" AS Col15, "attribute_attribute"."available_in_grid" AS Col16 FROM "attribute_attribute" LEFT OUTER JOIN "attribute_attributeproduct" ON ("attribute_attribute"."id" = "attribute_attributeproduct"."attribute_id") LEFT OUTER JOIN "attribute_attributevariant" ON ("attribute_attribute"."id" = "att

#3577107ms
SELECT "product_productvariant"."id", "product_productvariant"."sort_order", "product_productvariant"."private_metadata", "product_productvariant"."metadata", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."product_id", "product_productvariant"."track_inventory", "product_productvariant"."is_preorder", "product_productvariant"."preorder_end_date", "product_productvariant"."preorder_global_threshold", "product_productvariant"."quantity_limit_per_customer", "product_productvariant"."created_at", "product_productvariant"."updated_at", "product_productvariant"."weight" FROM "product_productvariant" WHERE ("product_productvariant"."is_preorder" AND "product_productvariant"."preorder_end_date" < '2022-06-19T18:24:27.297199+00:00'::timestamptz) ORDER BY "product_productvariant"."sort_order" ASC, "product_productvariant"."sku" ASC

# 3574246ms
SELECT "product_productvariant"."id", "product_productvariant"."sort_order", "product_productvariant"."private_metadata", "product_productvariant"."metadata", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."product_id", "product_productvariant"."track_inventory", "product_productvariant"."is_preorder", "product_productvariant"."preorder_end_date", "product_productvariant"."preorder_global_threshold", "product_productvariant"."quantity_limit_per_customer", "product_productvariant"."created_at", "product_productvariant"."updated_at", "product_productvariant"."weight" FROM "product_productvariant" WHERE ("product_productvariant"."is_preorder" AND "product_productvariant"."preorder_end_date" < '2022-06-19T18:24:27.297199+00:00'::timestamptz) ORDER BY "product_productvariant"."sort_order" ASC, "product_productvariant"."sku" ASC

# 2681528ms
SELECT COUNT(*) FROM (SELECT DISTINCT "attribute_attribute"."id" AS Col1, "attribute_attribute"."private_metadata" AS Col2, "attribute_attribute"."metadata" AS Col3, "attribute_attribute"."slug" AS Col4, "attribute_attribute"."name" AS Col5, "attribute_attribute"."type" AS Col6, "attribute_attribute"."input_type" AS Col7, "attribute_attribute"."entity_type" AS Col8, "attribute_attribute"."unit" AS Col9, "attribute_attribute"."value_required" AS Col10, "attribute_attribute"."is_variant_only" AS Col11, "attribute_attribute"."visible_in_storefront" AS Col12, "attribute_attribute"."filterable_in_storefront" AS Col13, "attribute_attribute"."filterable_in_dashboard" AS Col14, "attribute_attribute"."storefront_search_position" AS Col15, "attribute_attribute"."available_in_grid" AS Col16 FROM "attribute_attribute" LEFT OUTER JOIN "attribute_attributeproduct" ON ("attribute_attribute"."id" = "attribute_attributeproduct"."attribute_id") LEFT OUTER JOIN "attribute_attributevariant" ON ("attribute_attribute"."id" = "att

# This table is even empty and query has been running for 44.364 minutes
SELECT "product_collectiontranslation"."id", "product_collectiontranslation"."language_code", "product_collectiontranslation"."seo_title", "product_collectiontranslation"."seo_description", "product_collectiontranslation"."collection_id", "product_collectiontranslation"."name", "product_collectiontranslation"."description" FROM "product_collectiontranslation" WHERE ("product_collectiontranslation"."collection_id" IN (17) AND "product_collectiontranslation"."language_code" IN ('en-us'))
```

> My advice would be to use a free cloud version of Saleor for development, to free up yourself from the DevOps burden.

I get what you are saying for the average user, but that is not the direction I am going to head in.  I am expanding my DevOps abilities by managing more projects in my clusters.  The saleor graphQL based API is the only one with queries getting wedged, two other React + Django projects I have rolled myself in the cluster with more traditional REST APIs and not running an ASGI worker under uvicorn are not experiencing any kind of long running queries against this managed database.   I wonder if converting it to uwsgi which is what my other services are running under would be any more reliable than the current configuration, I might experiment with that or to see if there are some auto-kill gunicorn workers after certain time period. 

I am also waiting to hear back from Digital Ocean Engineering because AFAIK I wasn't having these sporadic network failures with the graphQL API earlier on with the project, and it really seems to be related to the long running queries, because if I manually terminate the queries, then I see in the logs of the saleor uvicorn pods that the workers receive the terminate query signal and output this to the logs:

```
{"asctime": "2022-06-20T02:29:03Z", "levelname": "ERROR", "lineno": 401, "message": "A query failed unexpectedly", "name": "saleor.graphql.errors.unhandled", "pathname": "/app/saleor/graphql/views.py", "process": 35, "threadName": "ThreadPoolExecutor-0_0", "exc_info": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py\", line 84, in _execute\n    return self.cursor.execute(sql, params)\npsycopg2.errors.UndefinedColumn: column payment_transactionitem.captured_value does not exist\nLINE 1: ...e_actions\", \"payment_transactionitem\".\"currency\", \"payment_t...\n                                                             ^\n\n\nThe above exception was the direct cause of the following exception:\n\nTraceback (most recent call last):\n  File \"/usr/local/lib/python3.9/site-packages/promise/dataloader.py\", line 261, in dispatch_queue_batch\n    batch_promise = loader.batch_load_fn(keys)\n  File \"/app/saleor/graphql/core/dataloaders.py\", line 45, in batch_load_fn\n    results = self.batch_load(keys)\n  File \"/app/saleor/graphql/order/dataloaders.py\", line 162, in batch_load\n    for transaction in transactions:\n  File \"/usr/local/lib/python3.9/site-packages/django/db/models/query.py\", line 280, in __iter__\n    self._fetch_all()\n  File \"/usr/local/lib/python3.9/site-packages/django/db/models/query.py\", line 1324, in _fetch_all\n    self._result_cache = list(self._iterable_class(self))\n  File \"/usr/local/lib/python3.9/site-packages/django/db/models/query.py\", line 51, in __iter__\n    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)\n  File \"/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py\", line 1175, in execute_sql\n    cursor.execute(sql, params)\n  File \"/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py\", line 66, in execute\n    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)\n  File \"/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py\", line 75, in _execute_with_wrappers\n    return executor(sql, params, many, context)\n  File \"/app/saleor/graphql/views.py\", line 52, in tracing_wrapper\n    return execute(sql, params, many, context)\n  File \"/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py\", line 84, in _execute\n    return self.cursor.execute(sql, params)\n  File \"/usr/local/lib/python3.9/site-packages/django/db/utils.py\", line 90, in __exit__\n    raise dj_exc_value.with_traceback(traceback) from exc_value\n  File \"/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py\", line 84, in _execute\n    return self.cursor.execute(sql, params)\ndjango.db.utils.ProgrammingError: column payment_transactionitem.captured_value does not exist\nLINE 1: ...e_actions\", \"payment_transactionitem\".\"currency\", \"payment_t...\n                                                             ^\n", "hostname": "mattscoinage-758d765c6c-9j7zv"}
[2022-06-20 02:33:01 +0000] [1] [WARNING] Worker with pid 35 was terminated due to signal 9
[2022-06-20 02:33:01 +0000] [166] [INFO] Booting worker with pid: 166
[2022-06-20 02:33:01 +0000] [166] [INFO] Started server process [166]
```

I am using the Dockerfile from within saleor to build the container and deploy it.   Backend DB is a managed postgresql instanace.

**Reply by matteius** [reply]
I'll add that I am now currently saleor on backend commit hash `ebc0e0e66`.  I know it would be better to run a tagged release version -- I am not even sure what the closest version that commit points to.  I tried running various tags all day Saturday/Sunday and decided that I would checkout to the exact commit hash of when I ran a successful ad campaign that experienced hundreds of users and did not have these performance issues.   The site is behaving slightly better but still experiencing issues under that time period.

**Reply by timuric** [reply]
@matteius we don't have a way to replicate this, and there were no other reports from the community, so we can't acknowledge this is an issue at this point. To me, it seems that this is something specific to your DO environment.

**Reply by matteius** [reply]
@timuric I think you are right about the long running queries being a DO environment thing.   I realize that during my debugging of the original outages, I created a connection pool.  The connection pool had not helped and so I tried changing it away from the default setting to be session based and it seems that is when the wedged query issue specifically started happening which made the sporadic issues way worse because it was tying up web workers.  When I changed it back to the default Transaction based pooling mode, the list of queries remains relatively short (not to say there aren't some long running ones) 
![image](https://user-images.githubusercontent.com/479892/174634179-56b7816a-b44c-4fdb-b039-92e69adebfd9.png)

Long running under this config:
`DECLARE "_django_curs_140365819463424_sync_17" NO SCROLL CURSOR WITH HOLD FOR SELECT "product_productmedia"."id", "product_productmedia"."sort_order", "product_productmedia"."product_id", "product_productmedia"."image", "product_productmedia"."ppoi", "product_productmedia"."alt", "product_productmedia"."type", "product_productmedia"."external_url", "product_productmedia"."oembed_data", "product_productmedia"."to_remove" FROM "product_productmedia" WHERE ("product_productmedia"."product_id" IN (78, 8, 31, 3, 23) AND NOT "product_productmedia"."to_remove") ORDER BY "product_productmedia"."sort_order" ASC, "product_productmedia"."id" ASC` has been running for 200+ seconds

`SELECT 1` running for 231449ms

`SELECT "page_page"."id", "page_page"."published_at", "page_page"."is_published", "page_page"."private_metadata", "page_page"."metadata", "page_page"."seo_title", "page_page"."seo_description", "page_page"."slug", "page_page"."title", "page_page"."page_type_id", "page_page"."content", "page_page"."created_at" FROM "page_page" WHERE "page_page"."id" IN (5, 3, 2, 4, 6)` running for 278816ms


Three though is better than all 24 workers being tied up this way, and I see where these queries do eventually finish.   I can vouch and say that the session based connection pool is definitely the wrong way to configure this service.  I am not sure why I still get the occasional wicked slow loading storefront page and a 500 server error, but this is way less sporadic than I was dealing with since Thursday.

Thanks for your help @timuric -- I may cautiously try to upgrade forward saleor backend this evening and deploy the newer storefront version I had been running for a couple weeks.  I am staring to see where there are some blog posts talking about the different versions ... maybe latest 3.4.x is the way to go?  Do you have a recommendation on what version to go with?

**Reply by matteius** [reply]
I marked this as technically answered.  I am still trying to get the to bottom of API failures like a random product collection or category fails to load on the page when everything else loaded:
![image](https://user-images.githubusercontent.com/479892/174637012-3d33514b-6ecd-4b3a-bc8b-8803b56d001f.png)

I am optimistic that upgrading saleor tonight might have some improvement with regards to this, but I am unsure.  One thing I know for sure is I way upped the worker count of the saleor backend so it shouldn't be any contention for resources with the current configuration and seeing no SQL queries tying things up.

**Reply by timuric** [reply]
As a rule of thumb, I would advise always picking the latest tagged release.

**Comment by matteius** [comment]
I discovered that even the preferred connection pool mode was causing me errors.  With the pool mode in the preferred `Transaction` mode I was getting sproadic graph QL errors.  Running the server locally against the same connetion pool yielded these same type of errors.  When I switched the connection string back to a regular non-pooled version, these errors went away:

```
21/Jun/2022 00:20:36] "POST /graphql/ HTTP/1.1" 200 599
[21/Jun/2022 00:20:36] "GET /static/images/placeholder255x255.png?w=1920 HTTP/1.1" 200 4710
ERROR saleor.graphql.errors.unhandled A query failed unexpectedly [PID:124032:Thread-8]
Traceback (most recent call last):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/promise/dataloader.py", line 261, in dispatch_queue_batch
    batch_promise = loader.batch_load_fn(keys)
  File "/home/matteius/mattscoinage/saleor/saleor/graphql/core/dataloaders.py", line 45, in batch_load_fn
    results = self.batch_load(keys)
  File "/home/matteius/mattscoinage/saleor/saleor/graphql/product/dataloaders/products.py", line 642, in batch_load
    for category in categories.iterator():
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/query.py", line 353, in _iterator
    yield from self._iterable_class(self, chunked_fetch=use_chunked_fetch, chunk_size=chunk_size)
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/query.py", line 68, in __iter__
    for row in compiler.results_iter(results):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1118, in apply_converters
    for row in map(list, rows):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1649, in cursor_iter
    cursor.close()
psycopg2.errors.InvalidCursorName: cursor "_django_curs_139953522603584_sync_1" does not exist

[21/Jun/2022 00:20:36] "POST /graphql/ HTTP/1.1" 200 2482
ERROR saleor.graphql.errors.unhandled A query failed unexpectedly [PID:124032:Thread-20]
Traceback (most recent call last):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.DuplicateCursor: cursor "_django_curs_139953522603584_sync_1" already exists


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/matteius/mattscoinage/saleor/saleor/graphql/views.py", line 53, in tracing_wrapper
    return execute(sql, params, many, context)
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: cursor "_django_curs_139953522603584_sync_1" already exists


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/promise/dataloader.py", line 261, in dispatch_queue_batch
    batch_promise = loader.batch_load_fn(keys)
  File "/home/matteius/mattscoinage/saleor/saleor/graphql/core/dataloaders.py", line 45, in batch_load_fn
    results = self.batch_load(keys)
  File "/home/matteius/mattscoinage/saleor/saleor/graphql/product/dataloaders/products.py", line 642, in batch_load
    for category in categories.iterator():
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/query.py", line 353, in _iterator
    yield from self._iterable_class(self, chunked_fetch=use_chunked_fetch, chunk_size=chunk_size)
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/query.py", line 51, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1178, in execute_sql
    cursor.close()
psycopg2.errors.InvalidCursorName: cursor "_django_curs_139953522603584_sync_1" does not exist

[21/Jun/2022 00:20:37] "POST /graphql/ HTTP/1.1" 200 4378
[21/Jun/2022 00:20:37] "POST /graphql/ HTTP/1.1" 200 619
[21/Jun/2022 00:20:37] "POST /graphql/ HTTP/1.1" 200 458
[21/Jun/2022 00:20:37] "POST /graphql/ HTTP/1.1" 200 182
[21/Jun/2022 00:20:37] "POST /graphql/ HTTP/1.1" 200 182
[21/Jun/2022 00:20:37] "POST /graphql/ HTTP/1.1" 200 182
ERROR saleor.graphql.errors.unhandled A query failed unexpectedly [PID:124032:Thread-21]
Traceback (most recent call last):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/utils.py", line 97, in inner
    return func(*args, **kwargs)
psycopg2.errors.InvalidCursorName: cursor "_django_curs_139952985732672_sync_1" does not exist


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/promise/dataloader.py", line 261, in dispatch_queue_batch
    batch_promise = loader.batch_load_fn(keys)
  File "/home/matteius/mattscoinage/saleor/saleor/graphql/core/dataloaders.py", line 45, in batch_load_fn
    results = self.batch_load(keys)
  File "/home/matteius/mattscoinage/saleor/saleor/graphql/product/dataloaders/products.py", line 642, in batch_load
    for category in categories.iterator():
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/query.py", line 353, in _iterator
    yield from self._iterable_class(self, chunked_fetch=use_chunked_fetch, chunk_size=chunk_size)
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/query.py", line 68, in __iter__
    for row in compiler.results_iter(results):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1118, in apply_converters
    for row in map(list, rows):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1646, in cursor_iter
    for rows in iter((lambda: cursor.fetchmany(itersize)), sentinel):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1646, in <lambda>
    for rows in iter((lambda: cursor.fetchmany(itersize)), sentinel):
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/utils.py", line 97, in inner
    return func(*args, **kwargs)
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/matteius/.virtualenvs/saleor/lib/python3.9/site-packages/django/db/utils.py", line 97, in inner
    return func(*args, **kwargs)
django.db.utils.OperationalError: cursor "_django_curs_139952985732672_sync_1" does not exist

[21/Jun/2022 00:20:38] "POST /graphql/ HTTP/1.1" 200 3572
[21/Jun/2022 00:20:38] "POST /graphql/ HTTP/1.1" 200 182
[21/Jun/2022 00:20:38] "POST /graphql/ HTTP/1.1" 200 182
[21/Jun/2022 00:20:38] "POST /graphql/ HTTP/1.1" 200 299
[21/Jun/2022 00:20:38] "POST /graphql/ HTTP/1.1" 200 932
[21/Jun/2022 00:20:38] "POST /graphql/ HTTP/1.1" 200 912
[21/Jun/2022 00:20:38] "POST /graphql/ HTTP/1.1" 200 463
```

**Reply by patrys** [reply]
The above errors indicate a pooler configured for per-statement routing (each individual SQL command sent to a different server) rather than transaction binding (all commands in the same transaction sent through the same connection). It can't work unless the entire transaction is targeting the same connection.

**Reply by matteius** [reply]
Thanks for that extra detail @patrys . I haven't gone back to see if I can get a version of connection pooling working, perhaps I will given this knowledge.  The root of my performance and sporadic outages issue ended up being this issue with proxy headers being broken for communication between services in the kubernetes cluster that do not route through the LB.  This issue seems specific to Digital Ocean and there is a fix described that worked for me here,  Reference: https://github.com/kubernetes/ingress-nginx/issues/3996

**Comment by matteius** [comment]
I think I finally got to the bottom of this tonight.  I upgraded my ingress-nginx-controller and continued to have this issue, I narrowed it down to the saleor API running in Docker behind ingress-nginx and the load balancer.  It turns out that this issue likely was the culprit and I followed one of the workarounds in it to get the broken headers to go away from the logs -- I never saw the broken headers in the logs until I first upgraded my ingress-nginx-controller.  Reference:  https://github.com/kubernetes/ingress-nginx/issues/3996

**Comment by ohuu** [comment]
I'm getting the exact same issue. Long running queries that never terminate. Can't use pooling because it breaks server side cursors and there's no documentation for turning server side cursors off so I'm completely stuck.

**Reply by matteius** [reply]
Well as I mentioned in my last post, I finally got to the bottom of it and included a link to the ingress-nginx report that helped me fix it.  Are you in Digital Ocean?

**Comment by ohuu** [comment]
I think our setup is quite different. I am using Digital Ocean docker (and nginx in a container) but I'm not using kubernetes and I'm not sure what the connection between nginx and postgres is. Maybe the error was the same for us but not the cause?

**Reply by matteius** [reply]
Trying to keep the issue discussion contained within a thread.   I bet the issue is still the same, and I recommend disabling all connection pooling while you get it figured out.  Just wondering, how are you using docker outside of kubernetes, and is that not kind of a pain to manage?  Key question:  Are you using a load balancer for your service?  The issue was that the react-storefront container was speaking directly to the saleor API container without going through the digital ocean load balancer.   I found you do need Proxy Protocol: Enabled for the load balancer, but that ingress-nginx doesn't handle the proxy protocol headers properly unless the request is routed through the load balancer (and doesn't talk directly to the backend service via nginx).   

For ingress-nginx there is an annotation that perhaps doesn't exist for regular nginx, that basically fixes the proxy protocol I believe by ensuring that request are routed through the hostname (which forces it through the LB which applies the proxy protocol headers appropriately).  
`service.beta.kubernetes.io/do-loadbalancer-hostname:`

What does your nginx config look like?  What you may want to do if you haven't already is assign your API a subdomain through the LB such as `api.yourdomain`  and then configure your instances to communicate directly through this DNS.  Somewhere along the line I believe your application has resolved its DNS and is communicating directly to the nginx server (without going through a LB) and loosing the proxy headers.   Have you looked for anything unusual in the nginx logs?
